src/sagas/nanoContract.js

```diff
@@ -167,33 +174,60 @@ export function* registerNanoContract({ payload }) {
 
 /**
  * Fetch history from Nano Contract wallet's API.
- * @param {string} ncId Nano Contract ID
- * @param {number} count Maximum quantity of history items
- * @param {string} after Transaction hash to start to get items
- * @param {Object} wallet Wallet instance from redux state
+ * @req {Object} req
+ * @param {string} req.ncId Nano Contract ID
+ * @param {number} req.count Maximum quantity of items to fetch
+ * @param {string} req.before Transaction hash to start to get newer items
+ * @param {string} req.after Transaction hash to start to get older items
+ * @param {Object} req.wallet Wallet instance from redux state
+ *
+ * @returns {{
+ *   history: {};
+ * }}
  *
  * @throws {Error} when request code is greater then 399 or when response's success is false
  */
-export async function fetchHistory(ncId, count, after, wallet) {
+export async function fetchHistory(req) {
+  const {
+    wallet,
+    ncId,
+    count,
+    after,
+    before,
+  } = req;
   /**
    * @type {RawNcTxHistoryResponse} response
    */
-  const response = await ncApi.getNanoContractHistory(ncId, count, after);
+  const response = await ncApi.getNanoContractHistory(
+    ncId,
+    count,
+    after || null,
+    before || null,
+  );
   const { success, history: rawHistory } = response;
 
   if (!success) {
     throw new Error('Failed to fetch nano contract history');
   }
 
-  const history = [];
-  for (const rawTx of rawHistory) {
+  /* TODO: Make it run concurrently while guaranting the order.
+  /* see https://github.com/HathorNetwork/hathor-wallet-mobile/issues/514
+   */
+  const historyNewestToOldest = new Array(rawHistory.length)
+  for (let idx = 0; idx < rawHistory.length; idx += 1) {
+    const rawTx = rawHistory[idx];
     const network = wallet.getNetworkObject();
     const caller = addressUtils.getAddressFromPubkey(rawTx.nc_pubkey, network).base58;
-    // XXX: Wallet Service Wallet doesn't implement isAddressMine.
-    // It means this method can't run under wallet-service without
-    // throwing an exception.
-    // eslint-disable-next-line no-await-in-loop
-    const isMine = await wallet.isAddressMine(caller);
+    // XXX: Wallet Service doesn't implement isAddressMine.
+    // See issue: https://github.com/HathorNetwork/hathor-wallet-lib/issues/732
+    // Default to `false` if using Wallet Service.
+    let isMine = false;
+    const useWalletService = consumeGenerator(isWalletServiceEnabled());
```

Comment:
Why do you need that?

Reply:
Can't you just use

`const useWalletService = yield call(isWalletServiceEnabled);`

?

Reply:
The function isn't a generator, but a promise. As we can't have async generators the only option to call one until its end is by consuming it. But in this case I'm interested in the result of it, not the yield values.